File: packages/node/src/Seguranca/Ataques/AtaqueBateriaEstatistica.js
'use strict';

const ResultadoAtaque = require('../ResultadoAtaque.js');
const SkipAtaqueException = require('../SkipAtaqueException.js');
const { randomBytes } = require('../Util.js');

/**
 * Bateria estatística no keystream (inspirada em NIST SP800-22): frequência
 * global (monobit), frequência por posição de bit, teste de runs e frequência
 * por bloco. Diferente do qui-quadrado de bytes, aqui o foco é o nível de BIT
 * e a estrutura de sequência - vieses que a contagem de bytes pode mascarar.
 *
 * Usa z-scores com limite conservador (4 desvios) em vez de p-valores exatos,
 * pra não depender de funções numéricas especiais.
 */
class AtaqueBateriaEstatistica {
  constructor({ tamanho = 16384, zLimite = 4.0 } = {}) {
    this.tamanho = tamanho;
    this.zLimite = zLimite;
  }

  nome() {
    return 'Bateria estatística de bits (monobit/runs/blocos)';
  }

  executar(alvo) {
    const ks = alvo.gerarKeystreamBruto(
      alvo.chaveDeTeste(),
      randomBytes(alvo.tamanhoIv()),
      'enc',
      this.tamanho,
    );
    if (ks === null) {
      throw new SkipAtaqueException('Alvo não expõe gerarKeystreamBruto.');
    }

    const bits = this.paraBits(ks);
    const n = bits.length;
    const problemas = [];
    const dados = {};

    // 1) Monobit: soma +1/-1
    let soma = 0;
    for (const b of bits) {
      soma += b === 1 ? 1 : -1;
    }
    const zMonobit = Math.abs(soma) / Math.sqrt(n);
    dados.monobit_z = zMonobit;
    if (zMonobit > this.zLimite) {
      problemas.push(`monobit z=${zMonobit.toFixed(2)}`);
    }

    // 2) Frequência por posição de bit
    const unsPorPos = new Array(8).fill(0);
    const contaPorPos = new Array(8).fill(0);
    for (let i = 0; i < ks.length; i++) {
      const byte = ks[i];
      for (let b = 0; b < 8; b++) {
        if (((byte >> b) & 1) === 1) {
          unsPorPos[b]++;
        }
        contaPorPos[b]++;
      }
    }
    const posViesadas = {};
    for (let b = 0; b < 8; b++) {
      const frac = unsPorPos[b] / contaPorPos[b];
      const z = Math.abs(frac - 0.5) / (0.5 / Math.sqrt(contaPorPos[b]));
      if (z > this.zLimite) {
        posViesadas[b] = frac;
      }
    }
    dados.bit_posicao_viesadas = posViesadas;
    if (Object.keys(posViesadas).length > 0) {
      problemas.push('viés por posição de bit: ' + Object.keys(posViesadas).join(', '));
    }

    // 3) Runs test
    const pi = bits.reduce((acc, b) => acc + b, 0) / n;
    if (Math.abs(pi - 0.5) < 2 / Math.sqrt(n)) {
      let runs = 1;
      for (let i = 1; i < n; i++) {
        if (bits[i] !== bits[i - 1]) {
          runs++;
        }
      }
      const esperado = 2 * n * pi * (1 - pi);
      const desvio = 2 * Math.sqrt(2 * n) * pi * (1 - pi);
      const zRuns = Math.abs(runs - esperado) / desvio;
      dados.runs_z = zRuns;
      if (zRuns > this.zLimite) {
        problemas.push(`runs z=${zRuns.toFixed(2)}`);
      }
    }

    // 4) Frequência por bloco (M = 128 bits)
    const m = 128;
    const numBlocos = Math.trunc(n / m);
    if (numBlocos > 0) {
      let chi = 0.0;
      for (let i = 0; i < numBlocos; i++) {
        let uns = 0;
        for (let j = 0; j < m; j++) {
          uns += bits[i * m + j];
        }
        chi += (uns / m - 0.5) ** 2;
      }
      chi *= 4 * m;
      const zBlocos = (chi - numBlocos) / Math.sqrt(2 * numBlocos);
      dados.blocos_chi2 = chi;
      dados.blocos_z = zBlocos;
      if (zBlocos > this.zLimite) {
        problemas.push(`frequência por bloco chi2=${chi.toFixed(1)}`);
      }
    }

    if (problemas.length > 0) {
      return new ResultadoAtaque(this.nome(), true, 'media', problemas.join('; '), dados);
    }

    return new ResultadoAtaque(
      this.nome(),
      false,
      'info',
      `monobit z=${zMonobit.toFixed(2)}, runs z=${(dados.runs_z ?? 0.0).toFixed(2)}, blocos chi2=${(dados.blocos_chi2 ?? 0.0).toFixed(1)} - todos abaixo do limite z=${this.zLimite.toFixed(0)}`,
      dados,
    );
  }

  paraBits(bytes) {
    const bits = [];
    for (let i = 0; i < bytes.length; i++) {
      const byte = bytes[i];
      for (let b = 7; b >= 0; b--) {
        bits.push((byte >> b) & 1);
      }
    }
    return bits;
  }
}

module.exports = AtaqueBateriaEstatistica;

File: packages/python/src/Seguranca/Ataques/AtaqueBateriaEstatistica.py
"""Bateria estatística no keystream (inspirada em NIST SP800-22): frequência
global (monobit), frequência por posição de bit, teste de runs e frequência
por bloco. Diferente do qui-quadrado de bytes, aqui o foco é o nível de BIT
e a estrutura de sequência - vieses que a contagem de bytes pode mascarar.

Usa z-scores com limite conservador (4 desvios) em vez de p-valores exatos,
pra não depender de funções numéricas especiais.
"""

import math

from ..ResultadoAtaque import ResultadoAtaque
from ..SkipAtaqueException import SkipAtaqueException
from ..Util import random_bytes


class AtaqueBateriaEstatistica:
    def __init__(self, tamanho: int = 16384, z_limite: float = 4.0):
        self.tamanho = tamanho
        self.z_limite = z_limite

    def nome(self) -> str:
        return "Bateria estatística de bits (monobit/runs/blocos)"

    def executar(self, alvo) -> ResultadoAtaque:
        ks = alvo.gerar_keystream_bruto(
            alvo.chave_de_teste().encode(), random_bytes(alvo.tamanho_iv()), "enc", self.tamanho
        )
        if ks is None:
            raise SkipAtaqueException("Alvo não expõe gerar_keystream_bruto.")

        bits = self._para_bits(ks)
        n = len(bits)
        problemas = []
        dados = {}

        # 1) Monobit: soma +1/-1
        soma = 0
        for b in bits:
            soma += 1 if b == 1 else -1
        z_monobit = abs(soma) / math.sqrt(n)
        dados["monobit_z"] = z_monobit
        if z_monobit > self.z_limite:
            problemas.append(f"monobit z={z_monobit:.2f}")

        # 2) Frequência por posição de bit
        uns_por_pos = [0] * 8
        conta_por_pos = [0] * 8
        for c in ks:
            byte = c
            for b in range(8):
                if ((byte >> b) & 1) == 1:
                    uns_por_pos[b] += 1
                conta_por_pos[b] += 1
        pos_viesadas = {}
        for b in range(8):
            uns = uns_por_pos[b]
            frac = uns / conta_por_pos[b]
            z = abs(frac - 0.5) / (0.5 / math.sqrt(conta_por_pos[b]))
            if z > self.z_limite:
                pos_viesadas[b] = frac
        dados["bit_posicao_viesadas"] = pos_viesadas
        if pos_viesadas:
            problemas.append("viés por posição de bit: " + ", ".join(str(k) for k in pos_viesadas.keys()))

        # 3) Runs test
        pi = sum(bits) / n
        if abs(pi - 0.5) < 2 / math.sqrt(n):
            runs = 1
            for i in range(1, n):
                if bits[i] != bits[i - 1]:
                    runs += 1
            esperado = 2 * n * pi * (1 - pi)
            desvio = 2 * math.sqrt(2 * n) * pi * (1 - pi)
            z_runs = abs(runs - esperado) / desvio
            dados["runs_z"] = z_runs
            if z_runs > self.z_limite:
                problemas.append(f"runs z={z_runs:.2f}")

        # 4) Frequência por bloco (M = 128 bits)
        m = 128
        num_blocos = n // m
        if num_blocos > 0:
            chi = 0.0
            for i in range(num_blocos):
                uns = 0
                for j in range(m):
                    uns += bits[i * m + j]
                chi += ((uns / m) - 0.5) ** 2
            chi *= 4 * m
            z_blocos = (chi - num_blocos) / math.sqrt(2 * num_blocos)
            dados["blocos_chi2"] = chi
            dados["blocos_z"] = z_blocos
            if z_blocos > self.z_limite:
                problemas.append(f"frequência por bloco chi2={chi:.1f}")

        if problemas:
            return ResultadoAtaque(self.nome(), True, "media", "; ".join(problemas), dados)

        return ResultadoAtaque(
            self.nome(),
            False,
            "info",
            (
                f"monobit z={z_monobit:.2f}, runs z={dados.get('runs_z', 0.0):.2f}, "
                f"blocos chi2={dados.get('blocos_chi2', 0.0):.1f} - todos abaixo do limite z={self.z_limite:.0f}"
            ),
            dados,
        )

    def _para_bits(self, bytes_: bytes) -> list:
        bits = []
        for i in range(len(bytes_)):
            byte = bytes_[i]
            for b in range(7, -1, -1):
                bits.append((byte >> b) & 1)
        return bits
